File: src/data/tradeDictionary.ts
import { InvoiceItem, PdfLanguage } from '../types';

export interface TermMapping {
  hindi: string[];
  hinglish: string[];
  english: string;
  category: 'electrical' | 'plumbing' | 'carpentry' | 'ac_appliance' | 'general';
  defaultRate?: number;
  defaultUnit?: string;
}

export const TRADE_MAPPINGS: TermMapping[] = [
  // Electrical
  {
    hindi: ['पंखा', 'पंखे', 'छत का पंखा'],
    hinglish: ['pankha', 'pankhe', 'fan', 'ceiling fan'],
    english: 'Ceiling Fan',
    category: 'electrical',
    defaultRate: 350,
    defaultUnit: 'nos',
  },
  {
    hindi: ['फिटिंग', 'लगाना', 'फिक्सिंग', 'इन्स्टॉलेशन'],
    hinglish: ['fitting', 'lagana', 'lagai', 'fixing', 'installation', 'fit'],
    english: 'Installation & Fitting',
    category: 'general',
    defaultRate: 250,
    defaultUnit: 'job',
  },
  {
    hindi: ['तार', 'वायर', 'केबल'],
    hinglish: ['taar', 'tar', 'wire', 'cable', 'wiring'],
    english: 'Electrical Wiring',
    category: 'electrical',
    defaultRate: 450,
    defaultUnit: 'mtr/coil',
  },
  {
    hindi: ['स्विच', 'स्विच बोर्ड', 'बोर्ड', 'प्लग'],
    hinglish: ['switch', 'switch board', 'switchboard', 'board', 'plug', 'socket'],
    english: 'Modular Switch Board',
    category: 'electrical',
    defaultRate: 180,
    defaultUnit: 'nos',
  },
  {
    hindi: ['एमसीबी', 'एमसीबी बॉक्स', 'कट आउट'],
    hinglish: ['mcb', 'mcb box', 'cutout', 'trip switch'],
    english: 'MCB Circuit Breaker',
    category: 'electrical',
    defaultRate: 450,
    defaultUnit: 'nos',
  },
  {
    hindi: ['लाइट', 'बल्ब', 'ट्यूबलाइट', 'एलईडी'],
    hinglish: ['light', 'bulb', 'tubelight', 'led', 'cfl'],
    english: 'LED Fixture & Tubelight',
    category: 'electrical',
    defaultRate: 120,
    defaultUnit: 'nos',
  },

  // Plumbing
  {
    hindi: ['पाइप', 'नल का पाइप', 'पीवीसी पाइप'],
    hinglish: ['pipe', 'pvc pipe', 'cpvc pipe', 'water pipe'],
    english: 'PVC / Copper Piping',
    category: 'plumbing',
    defaultRate: 350,
    defaultUnit: 'ft',
  },
  {
    hindi: ['नल', 'टोटी', 'नलका'],
    hinglish: ['nal', 'toti', 'tap', 'bib cock', 'water tap'],
    english: 'Brass / Chrome Water Tap',
    category: 'plumbing',
    defaultRate: 280,
    defaultUnit: 'nos',
  },
  {
    hindi: ['गीज़र', 'गीजर फिटिंग', 'वाटर हीटर'],
    hinglish: ['geyser', 'geezer', 'water heater', 'geyser fitting'],
    english: 'Water Geyser Installation',
    category: 'plumbing',
    defaultRate: 650,
    defaultUnit: 'job',
  },
  {
    hindi: ['मोटर', 'सबमर्सिबल', 'पंप'],
    hinglish: ['motor', 'submersible', 'water pump', 'tullu motor'],
    english: 'Water Pump Repair & Service',
    category: 'plumbing',
    defaultRate: 500,
    defaultUnit: 'job',
  },
  {
    hindi: ['लीकेज', 'रिसाव', 'टपकना'],
    hinglish: ['leakage', 'leaking', 'tap tapakna', 'seepage'],
    english: 'Leakage Detection & Seal Repair',
    category: 'plumbing',
    defaultRate: 300,
    defaultUnit: 'job',
  },

  // AC & Appliances
  {
    hindi: ['एसी सर्विस', 'एसी गैस', 'एसी रिपेयर'],
    hinglish: ['ac service', 'ac repair', 'ac gas', 'cooling repair'],
    english: 'AC Deep Clean Service & Gas Charge',
    category: 'ac_appliance',
    defaultRate: 850,
    defaultUnit: 'nos',
  },

  // Carpentry
  {
    hindi: ['कब्जा', 'हिंज'],
    hinglish: ['kabja', 'hinge', 'door hinge'],
    english: 'Heavy-Duty Door Hinges',
    category: 'carpentry',
    defaultRate: 150,
    defaultUnit: 'pair',
  },
  {
    hindi: ['ताला', 'हैंडल लॉक', 'लॉक'],
    hinglish: ['tala', 'lock', 'handle lock', 'door lock'],
    english: 'Mortise Handle Lock Fitting',
    category: 'carpentry',
    defaultRate: 400,
    defaultUnit: 'nos',
  },
];

/**
 * Normalizes input text according to the selected language mode.
 */
export function normalizeTradeText(text: string, language: PdfLanguage): string {
  if (!text) return '';
  const lower = text.trim().toLowerCase();

  for (const mapping of TRADE_MAPPINGS) {
    // Check Hindi matches
    for (const h of mapping.hindi) {
      if (lower.includes(h.toLowerCase())) {
        return language === 'ENGLISH' ? mapping.english : mapping.hinglish[0];
      }
    }
    // Check Hinglish matches
    for (const hg of mapping.hinglish) {
      if (lower.includes(hg.toLowerCase())) {
        return language === 'ENGLISH' ? mapping.english : mapping.hinglish[0];
      }
    }
  }

  // If already in English or custom text, capitalize nicely
  return text.charAt(0).toUpperCase() + text.slice(1);
}

/**
 * Parses spoken Hinglish voice input into structured invoice items.
 * Example input: "1 pankha fitting 350 aur 2 switch board 300 taar 200"
 */
export function parseVoiceInput(spokenTranscript: string, language: PdfLanguage): InvoiceItem[] {
  if (!spokenTranscript.trim()) return [];

  // Split clauses by separators like 'aur', 'and', 'bhi', ',', '+'
  const segments = spokenTranscript
    .toLowerCase()
    .split(/\b(?:aur|and|plus|bhi|\+|,)\b/gi)
    .map(s => s.trim())
    .filter(Boolean);

  const items: InvoiceItem[] = [];

  for (let idx = 0; idx < segments.length; idx++) {
    const segment = segments[idx];
    
    // Extract numbers: quantity (first number if 1-2 digits before item name) and rate/amount
    const numbers = (segment.match(/\d+/g) || []).map(Number);
    let qty = 1;
    let rate = 0;

    if (numbers.length === 1) {
      // Typically rate if > 50, otherwise qty
      if (numbers[0] > 50) {
        rate = numbers[0];
      } else {
        qty = numbers[0];
      }
    } else if (numbers.length >= 2) {
      qty = numbers[0];
      rate = numbers[1];
    }

    // Identify matching item
    let matchedMapping: TermMapping | null = null;
    let matchedSubstr = '';

    for (const mapping of TRADE_MAPPINGS) {
      const allKeywords = [...mapping.hinglish, ...mapping.hindi];
      for (const kw of allKeywords) {
        if (segment.includes(kw.toLowerCase())) {
          matchedMapping = mapping;
          matchedSubstr = kw;
          break;
        }
      }
      if (matchedMapping) break;
    }

    if (matchedMapping) {
      if (rate === 0 && matchedMapping.defaultRate) {
        rate = matchedMapping.defaultRate;
      }
      const englishName = matchedMapping.english;
      const hinglishName = matchedMapping.hinglish[0].toUpperCase() + ' ' + (matchedMapping.category === 'electrical' ? 'Ka Kaam' : '');
      const selectedName = language === 'ENGLISH' ? englishName : hinglishName;

      items.push({
        id: 'item_' + Date.now() + '_' + idx,
        originalText: segment,
        name: selectedName,
        normalizedEnglishName: englishName,
        hinglishName: hinglishName,
        quantity: qty,
        unit: matchedMapping.defaultUnit || 'nos',
        rate: rate || 250,
        amount: qty * (rate || 250),
      });
    } else {
      // Fallback for custom terms
      const cleanName = segment.replace(/\d+/g, '').replace(/ka|ki|ke|fitting|repair|service/gi, '').trim();
      const title = cleanName ? cleanName.charAt(0).toUpperCase() + cleanName.slice(1) : 'Service Work';
      const actualRate = rate > 0 ? rate : 300;
      items.push({
        id: 'item_' + Date.now() + '_' + idx,
        originalText: segment,
        name: title,
        normalizedEnglishName: title,
        hinglishName: title,
        quantity: qty,
        unit: 'nos',
        rate: actualRate,
        amount: qty * actualRate,
      });
    }
  }

  return items;
}

File: android/app/src/main/java/com/bolobill/app/util/TradeDictionary.kt
package com.bolobill.app.util

import java.util.Locale
import java.util.regex.Pattern

object TradeDictionary {
    enum class LanguageMode { HINGLISH, ENGLISH }

    data class TradeItemDefinition(
        val devanagariKeywords: List<String>,
        val hinglishKeywords: List<String>,
        val englishProfessional: String,
        val hinglishColloquial: String,
        val defaultRate: Double,
        val standardUnit: String,
        val tradeCategory: String
    )

    private val DICTIONARY = listOf(
        // ==================== ELECTRICAL TRADES ====================
        TradeItemDefinition(
            devanagariKeywords = listOf("पंखा", "पंखे", "सीलिंग फैन", "छत का पंखा"),
            hinglishKeywords = listOf("pankha", "pankhe", "fan", "ceiling fan"),
            englishProfessional = "Ceiling Fan",
            hinglishColloquial = "Ceiling Fan Fitting",
            defaultRate = 350.0,
            standardUnit = "nos",
            tradeCategory = "Electrical"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("तार", "वायर", "वायरिंग", "केबल"),
            hinglishKeywords = listOf("taar", "tar", "wire", "cable", "wiring"),
            englishProfessional = "Electrical Wiring",
            hinglishColloquial = "Taar Wiring",
            defaultRate = 450.0,
            standardUnit = "coil",
            tradeCategory = "Electrical"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("स्विच", "बोर्ड", "प्लग", "सॉकेट", "स्विच बोर्ड"),
            hinglishKeywords = listOf("switch", "switch board", "switchboard", "board", "plug", "socket"),
            englishProfessional = "Modular Switch Board",
            hinglishColloquial = "Switch Board Fitting",
            defaultRate = 180.0,
            standardUnit = "nos",
            tradeCategory = "Electrical"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("एमसीबी", "एमसीबी बॉक्स", "कट आउट", "ट्रिप स्विच"),
            hinglishKeywords = listOf("mcb", "mcb box", "cutout", "trip switch"),
            englishProfessional = "MCB Circuit Breaker",
            hinglishColloquial = "MCB Box Fitting",
            defaultRate = 450.0,
            standardUnit = "nos",
            tradeCategory = "Electrical"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("लाइट", "बल्ब", "ट्यूबलाइट", "एलईडी", "होल्डर"),
            hinglishKeywords = listOf("light", "bulb", "tubelight", "led", "cfl", "holder"),
            englishProfessional = "LED Fixture & Tubelight",
            hinglishColloquial = "Light Fitting",
            defaultRate = 150.0,
            standardUnit = "nos",
            tradeCategory = "Electrical"
        ),

        // ==================== PLUMBING TRADES ====================
        TradeItemDefinition(
            devanagariKeywords = listOf("नल", "टोटी", "टैप", "बिब कॉक"),
            hinglishKeywords = listOf("nal", "toti", "tap", "bib cock", "water tap"),
            englishProfessional = "Brass / Chrome Water Tap",
            hinglishColloquial = "Tap Fitting",
            defaultRate = 250.0,
            standardUnit = "nos",
            tradeCategory = "Plumbing"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("पाइप", "प्लंबिंग", "पीवीसी पाइप", "सीपीवीसी पाइप", "वाटर पाइप"),
            hinglishKeywords = listOf("pipe", "pvc pipe", "cpvc pipe", "water pipe", "plumbing pipe"),
            englishProfessional = "PVC / Copper Piping",
            hinglishColloquial = "Piping Line Work",
            defaultRate = 350.0,
            standardUnit = "ft",
            tradeCategory = "Plumbing"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("टंकी", "वाटर टैंक", "फ्लश", "सिस्टर्न"),
            hinglishKeywords = listOf("tanki", "tank", "water tank", "flush", "cistern"),
            englishProfessional = "Water Tank & Flush Repair",
            hinglishColloquial = "Tanki Repair Work",
            defaultRate = 650.0,
            standardUnit = "job",
            tradeCategory = "Plumbing"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("बेसिन", "वाश बेसिन", "सिंक", "नाली"),
            hinglishKeywords = listOf("basin", "wash basin", "washbasin", "sink", "drain", "nali"),
            englishProfessional = "Wash Basin & Sink Fitting",
            hinglishColloquial = "Basin Fitting Work",
            defaultRate = 400.0,
            standardUnit = "nos",
            tradeCategory = "Plumbing"
        ),

        // ==================== CARPENTRY TRADES ====================
        TradeItemDefinition(
            devanagariKeywords = listOf("दरवाजा", "किवाड़", "डोर", "चौखट"),
            hinglishKeywords = listOf("darwaza", "darwaja", "door", "chaukhat", "kivaad"),
            englishProfessional = "Door Fitting & Alignment",
            hinglishColloquial = "Darwaza Fitting",
            defaultRate = 500.0,
            standardUnit = "nos",
            tradeCategory = "Carpentry"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("अलमारी", "वार्डरोब", "कपाट", "कबर्ड"),
            hinglishKeywords = listOf("almari", "almirah", "wardrobe", "cupboard", "kapat"),
            englishProfessional = "Wooden Wardrobe Work",
            hinglishColloquial = "Almari Work",
            defaultRate = 800.0,
            standardUnit = "job",
            tradeCategory = "Carpentry"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("ताला", "लॉक", "हैंडल", "कब्ज़ा"),
            hinglishKeywords = listOf("tala", "lock", "handle", "kabza", "hinge", "hinges"),
            englishProfessional = "Lock & Hinges Fitting",
            hinglishColloquial = "Tala Kabza Fitting",
            defaultRate = 250.0,
            standardUnit = "nos",
            tradeCategory = "Carpentry"
        ),
        TradeItemDefinition(
            devanagariKeywords = listOf("प्लाई", "प्लाईवुड", "फर्नीचर", "टेबल", "कुर्सी"),
            hinglishKeywords = listOf("ply", "plywood", "furniture", "table", "kursi"),
            englishProfessional = "Plywood & Furniture Repair",
            hinglishColloquial = "Furniture Work",
            defaultRate = 450.0,
            standardUnit = "job",
            tradeCategory = "Carpentry"
        ),

        // ==================== GENERAL LABOUR ====================
        TradeItemDefinition(
            devanagariKeywords = listOf("फिटिंग", "लगाना", "फिक्सिंग", "मरम्मत"),
            hinglishKeywords = listOf("fitting", "lagana", "lagai", "fixing", "installation", "marammat"),
            englishProfessional = "Installation & Fitting",
            hinglishColloquial = "Fitting Work",
            defaultRate = 250.0,
            standardUnit = "job",
            tradeCategory = "General"
        )
    )

    fun normalizeTerm(rawText: String, targetLanguage: LanguageMode): String {
        val clean = rawText.trim().lowercase(Locale.ROOT)
        for (item in DICTIONARY) {
            if (item.hinglishKeywords.any { clean.contains(it) } || 
                item.devanagariKeywords.any { clean.contains(it) }) {
                return if (targetLanguage == LanguageMode.ENGLISH) item.englishProfessional else item.hinglishColloquial
            }
        }
        return rawText.replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.ROOT) else it.toString() }
    }

    fun parseSpokenUtterance(spokenText: String, targetLanguage: LanguageMode): List<ParsedItem> {
        val result = mutableListOf<ParsedItem>()
        val segments = spokenText.split(Regex("""(?i)\b(aur|and|plus|bhi|va|tatha|,)\b"""))
        
        for (seg in segments) {
            val trimmed = seg.trim()
            if (trimmed.isEmpty()) continue

            val matcher = Pattern.compile("""\d+""").matcher(trimmed)
            val numbers = mutableListOf<Int>()
            while (matcher.find()) {
                matcher.group().toIntOrNull()?.let { numbers.add(it) }
            }

            var qty = 1
            var rate = 0.0
            if (numbers.size == 1) {
                if (numbers[0] > 50) rate = numbers[0].toDouble() else qty = numbers[0]
            } else if (numbers.size >= 2) {
                qty = numbers[0]
                rate = numbers[1].toDouble()
            }

            val lower = trimmed.lowercase(Locale.ROOT)
            val matchedDef = DICTIONARY.firstOrNull { item ->
                item.hinglishKeywords.any { lower.contains(it) } ||
                item.devanagariKeywords.any { lower.contains(it) }
            }

            if (matchedDef != null) {
                if (rate == 0.0) rate = matchedDef.defaultRate
                val itemName = if (targetLanguage == LanguageMode.ENGLISH) matchedDef.englishProfessional else matchedDef.hinglishColloquial
                result.add(ParsedItem(itemName, qty, matchedDef.standardUnit, rate, qty * rate))
            } else {
                val cleanedName = trimmed.replace(Regex("""\d+"""), "").trim()
                val finalName = cleanedName.ifEmpty { "Service Work" }
                val finalRate = if (rate > 0.0) rate else 250.0
                result.add(ParsedItem(finalName, qty, "nos", finalRate, qty * finalRate))
            }
        }
        return result
    }

    data class ParsedItem(
        val name: String,
        val quantity: Int,
        val unit: String,
        val rate: Double,
        val amount: Double
    )
}
